serve/mlc_serve/model/tvm_model.py

```diff
@@ -37,21 +40,83 @@
 LOG = structlog.stdlib.get_logger(__name__)
 
 
-def load_disco_module(artifact_path, lib_path, num_shards):
-    sess = di.ProcessSession(num_workers=num_shards, entrypoint="tvm.exec.disco_worker")
-    devices = range(num_shards)
-    sess.init_ccl("nccl", *devices)
-    module = sess.load_vm_module(lib_path)
+def load_disco_weights_through_shard_loader(sess, param_path, lib_path):
+    # TODO(Lunderberg): Use a callback for get_item/set_item so these
+    # only need to be defined on the transform_params path.
+    worker_device = sess.get_global_func("runtime.disco.device")()
+    sess.import_python_module("mlc_serve.model.lazy_safetensor")
+    sess.get_global_func("mlc_serve.model.define_safetensors_get_item")(
+        ".", "", worker_device
+    )
+
+    metadata_path = param_path.joinpath("ndarray-cache.json")
+    module = sess.load_vm_module(lib_path.as_posix())
 
     loader_create = sess.get_global_func("runtime.disco.ShardLoader")
-    metadata_path = os.path.join(artifact_path, "params", "ndarray-cache.json")
-    with open(metadata_path, "r", encoding="utf-8") as f:
+    with metadata_path.open(encoding="utf-8") as f:
         ndarray_cache_metadata = f.read()
 
-    loader = loader_create(metadata_path, ndarray_cache_metadata, "", module)
+    loader = loader_create(metadata_path.as_posix(), ndarray_cache_metadata, "", module)
     loader_load = sess.get_global_func("runtime.disco.ShardLoaderLoadAllPresharded")
     params = loader_load(loader)
 
+    return module, params
+
+
+def load_disco_weights_through_transform_params(sess, orig_param_path, lib_path):
+    worker_id = sess.get_global_func("runtime.disco.worker_id")()
+    worker_device = sess.get_global_func("runtime.disco.device")()
+
+    @tvm.register_func("get_item", override=True)
+    def _unused_get_item():
+        pass
+
+    vm = relax.VirtualMachine(tvm.runtime.load_module(lib_path.as_posix()), tvm.cpu())
+    param_names = vm["get_original_param_names"]()
+
+    sess.import_python_module("mlc_serve.model.lazy_safetensor")
+    sess.get_global_func("mlc_serve.model.define_safetensors_get_item")(
+        orig_param_path.as_posix(), "\n".join(param_names), worker_device
+    )
+
+    module = sess.load_vm_module(lib_path.as_posix())
+    params = module["transform_params"](worker_id)
```

Comment:
That sounds reasonable for the short-term, and is a good workaround for the memory profiler limitations.  For the long-term, I'm wondering if we should make a transform that generates a memory footprint calculator for each function.  This transform would determine the liveset of variables at each step in the Relax compute graph, in terms of any symbolic shapes in the function parameters, and would generate a function that returns the maximum footprint across all livesets in a function.

What I'm picturing for the output:

```python
@R.function
def max_bytes_memory(
    func_name: str,
    vocab_size: R.Prim(value="tir_vocab_size"),
    seq_len: R.Prim(value="tir_seq_len"),
) -> R.Prim("int64"):
    tir_vocab_size = T.int64()
    tir_seq_len = T.int64()

    if func_name == "prefill":
        output = R.prim_value(...)
    elif func_name == "decode":
        output = R.prim_value(...)
    else:
        R.assert_op(
            False,
            func_name,
            format="Function name must in ['prefill','decode'], but was {}",
        )
        output = R.prim_value(-1)
    return output
```

That way, instead of needing to perform a memory-profiling call with a specific setup for the memory pool, we could call the generated function and use the precomputed memory footprint.

Reply:
yeah, as long as this approach computes the same value as the current dynamic one, I agree that we should switch to this approach. Feel free to work on it if you are interested (might be necessary for LoRA as discussed).